internal/commands/crypto.go

```diff
@@ -0,0 +1,442 @@
+		Args:    cobra.NoArgs,
+	}
+
+	cmd.AddCommand(
+		newCryptoPairRSACmd(),
+		newCryptoPairECDSACmd(),
+		newCryptoPairEd25519Cmd(),
+	)
+
+	return cmd
+}
+
+func newCryptoPairRSACmd() (cmd *cobra.Command) {
+	cmd = &cobra.Command{
+		Use:     cmdUseRSA,
+		Short:   cmdAutheliaCryptoPairRSAShort,
+		Long:    cmdAutheliaCryptoPairRSALong,
+		Example: cmdAutheliaCryptoPairRSAExample,
+		Args:    cobra.NoArgs,
+		RunE:    cryptoGenerateRunE,
+	}
+
+	cmd.AddCommand(newCryptoGenerateCmd(cmdUsePair, cmdUseRSA))
+
+	return cmd
+}
+
+func newCryptoPairECDSACmd() (cmd *cobra.Command) {
+	cmd = &cobra.Command{
+		Use:     cmdUseECDSA,
+		Short:   cmdAutheliaCryptoPairECDSAShort,
+		Long:    cmdAutheliaCryptoPairECDSALong,
+		Example: cmdAutheliaCryptoPairECDSAExample,
+		Args:    cobra.NoArgs,
+		RunE:    cryptoGenerateRunE,
+	}
+
+	cmd.AddCommand(newCryptoGenerateCmd(cmdUsePair, cmdUseECDSA))
+
+	return cmd
+}
+
+func newCryptoPairEd25519Cmd() (cmd *cobra.Command) {
+	cmd = &cobra.Command{
+		Use:     cmdUseEd25519,
+		Short:   cmdAutheliaCryptoPairEd25519Short,
+		Long:    cmdAutheliaCryptoPairEd25519Long,
+		Example: cmdAutheliaCryptoPairEd25519Example,
+		Args:    cobra.NoArgs,
+		RunE:    cryptoGenerateRunE,
+	}
+
+	cmd.AddCommand(newCryptoGenerateCmd(cmdUsePair, cmdUseEd25519))
+
+	return cmd
+}
+
+func cryptoGenerateRunE(cmd *cobra.Command, args []string) (err error) {
+	var (
+		privateKey interface{}
+	)
+
+	if privateKey, err = cryptoGenPrivateKeyFromCmd(cmd); err != nil {
+		return err
+	}
+
+	if cmd.Parent().Parent().Use == cmdUseCertificate {
+		return cryptoCertificateGenerateRunE(cmd, args, privateKey)
+	}
+
+	return cryptoPairGenerateRunE(cmd, args, privateKey)
+}
+
+func cryptoPairGenerateRunE(cmd *cobra.Command, _ []string, privateKey interface{}) (err error) {
+	var (
+		privateKeyPath, publicKeyPath string
+		pkcs8                         bool
+	)
+
+	if pkcs8, err = cmd.Flags().GetBool(cmdFlagNamePKCS8); err != nil {
+		return err
+	}
+
+	if privateKeyPath, publicKeyPath, err = cryptoGetWritePathsFromCmd(cmd); err != nil {
+		return err
+	}
+
+	b := strings.Builder{}
+
+	b.WriteString("Generating key pair\n\n")
+
+	switch k := privateKey.(type) {
+	case *rsa.PrivateKey:
+		b.WriteString(fmt.Sprintf("\tAlgorithm: RSA-%d %d bits\n\n", k.Size(), k.N.BitLen()))
+	case *ecdsa.PrivateKey:
+		b.WriteString(fmt.Sprintf("\tAlgorithm: ECDSA Curve %s\n\n", k.Curve.Params().Name))
+	case ed25519.PrivateKey:
+		b.WriteString("\tAlgorithm: Ed25519\n\n")
+	}
+
+	b.WriteString("Output Paths:\n")
+	b.WriteString(fmt.Sprintf("\tPrivate Key: %s\n", privateKeyPath))
+	b.WriteString(fmt.Sprintf("\tPublic Key: %s\n\n", publicKeyPath))
+
+	fmt.Print(b.String())
+
+	b.Reset()
+
+	if err = utils.WriteKeyToPEM(privateKey, privateKeyPath, pkcs8); err != nil {
+		return err
+	}
+
+	var publicKey interface{}
+
+	if publicKey = utils.PublicKeyFromPrivateKey(privateKey); publicKey == nil {
+		return fmt.Errorf("failed to obtain public key from private key")
+	}
```

Comment:
Is this testable? Could we simulate this by pointing it to a public key instead (or will that result in a completely different error)?

Reply:
This is tested in utils, however I'm not sure how we would test it in this specific case as the value sent here should always be a valid private key. It's always generated as part of the command, and an error should catch it before now.. maybe we just let this panic if it occurs?